cpp/pipelines/add_authority_external_ref.cc

```diff
@@ -130,19 +181,164 @@ void ParseGndWikidataMappingFile(std::string filename,
         LOG_ERROR("input or output files could not be opened");
 }
 
+struct GNDStructure {
+    std::string gnd_id, wikidata_personal_entity_id, wikipedia_personal_address;
+};
+
+bool IsThisCloseBracketForId(const std::string &url) {
+    int url_length = url.length();
+    std::string is_about(url.substr(url_length - 5, 5));
+    if (is_about.compare("about") == 0)
+        return true;
+
+    return false;
+}
+
+bool DoesItMatch(const std::string &url_based, const std::string &url_comp) {
+    const int base_string_lenght = url_based.length();
+    const std::string sub_string_url_comp = url_comp.substr(0, base_string_lenght);
+    if (url_based.compare(sub_string_url_comp) == 0)
+        return true;
+
+    return false;
+}
+
+bool GenerateGNDAuthorityExternalRef(char *argv[]) {
+    const auto load_file_start(std::chrono::high_resolution_clock::now());
+    std::ifstream input_file(argv[2]);
+
+    if (!input_file.is_open()) {
+        LOG_ERROR("can't open input file");
+        return false;
+    }
+
+    const auto csv_file(FileUtil::OpenOutputFileOrDie(argv[3]));
+
+
+    const std::string dnb_address("https://d-nb.info/gnd/");
+    const std::string wikidata_address("http://www.wikidata.org/entity/");
+    const std::string wikipedia_address("https://de.wikipedia.org/wiki/");
+    std::string gnd_id;
+    bool is_start_group = false;
+    GNDStructure gnd_data;
+    int top_level_number(-1), second_level_number(-1), total_numbers_of_gnd_id_generated(0), total_line_parsed(0),
+        total_number_of_wikidata(0), total_number_of_wikipedia(0);
+
+    const int dnb_add_str_lenght(dnb_address.length()), wikidata_address_str_lenght(wikidata_address.length());
+
+    std::string line, id_annotaton(""), second_element_of_array;
+    nlohmann::json line_parsed;
+    std::string gnd_id_temp_string;
+    std::string wikidata_temp_string;
+    std::string tmp_id;
+
 
-int Main(int argc, char* argv[]) {
-    if (argc != 4)
+    while (std::getline(input_file, line)) {
+        line_parsed = nlohmann::json::parse(line);
```

Comment:
The whole `GenerateGNDAuthorityExternalRef` function is very long. Maybe you can split it up a bit, e.g. move the parsing logic of a single line in the while loop into a separate function?

Reply:
And it seems to be very specific to the lines in the actual format => i think it would be good if you could show me the original / prepared file on your machine so i can give good feedback about it.